ykrt/src/compile/j2/opt/known_bits.rs

```diff
@@ -0,0 +1,474 @@
+/// https://pypy.org/posts/2024/08/toy-knownbits.html#the-knownbits-abstract-domain
+///
+/// In short:
+/// one     unknown     knownbit
+/// 0       0           0
+/// 1       0           1
+/// 0       1           ?
+/// 1       1           contradiction
+///
+/// To ensure monotonicity, state always goes upwards in this lattice.
+///           contradiction
+///             /      \
+///             1      0
+///             \      /
+///                ?
+#[derive(Clone)]
+pub struct KnownBitValue {
+    ones: ArbBitInt,
+    pub unknowns: ArbBitInt,
+}
+
+/// Full credits to the PyPy blog post for some of the functions here:
+/// https://pypy.org/posts/2024/08/toy-knownbits.html#the-knownbits-abstract-domain
+impl KnownBitValue {
+    fn from_constant(num: &ArbBitInt) -> Self {
+        let bitw = num.bitw();
+        KnownBitValue {
+            ones: num.clone(),
+            unknowns: ArbBitInt::from_u64(bitw, 0),
+        }
+    }
+
+    pub fn unknown(bitw: u32) -> Self {
+        KnownBitValue {
+            ones: ArbBitInt::from_u64(bitw, 0),
+            unknowns: ArbBitInt::from_u64(bitw, !0u64),
+        }
+    }
+
+    fn as_int(&self) -> ArbBitInt {
+        assert!(self.all_known());
+        self.ones.clone()
+    }
+
+    fn all_known(&self) -> bool {
+        self.unknowns.count_ones() == 0
+    }
+
+    /// Return an integer where the known bits are set
+    fn knowns(&self) -> ArbBitInt {
+        self.unknowns.bitneg()
+    }
+
+    /// Returns an integer where the places that are known zeros have a bit set
+    fn zeroes(&self) -> ArbBitInt {
+        self.knowns().bitand(&self.ones.bitneg())
+    }
+
+    fn known_ones(&self) -> ArbBitInt {
+        self.knowns().bitand(&self.ones)
+    }
+
+    fn and(&self, other: &KnownBitValue) -> KnownBitValue {
+        assert_eq!(self.ones.bitw(), self.unknowns.bitw());
+        assert_eq!(self.ones.bitw(), other.ones.bitw());
+        assert_eq!(self.unknowns.bitw(), other.unknowns.bitw());
+        let set_ones = self.ones.bitand(&other.ones);
+        let set_zeroes = self.zeroes().bitor(&other.zeroes());
+        let unknowns = self
+            .unknowns
+            .bitor(&other.unknowns)
+            .bitand(&set_zeroes.bitneg());
+        KnownBitValue {
+            ones: set_ones,
+            unknowns,
+        }
+    }
+
+    fn or(&self, other: &KnownBitValue) -> KnownBitValue {
+        assert_eq!(self.ones.bitw(), self.unknowns.bitw());
+        assert_eq!(self.ones.bitw(), other.ones.bitw());
+        assert_eq!(self.unknowns.bitw(), other.unknowns.bitw());
+        let set_ones = self.ones.bitor(&other.ones);
+        let unknowns = self
+            .unknowns
+            .bitor(&other.unknowns)
+            .bitand(&set_ones.bitneg());
+        KnownBitValue {
+            ones: set_ones,
+            unknowns,
+        }
+    }
+}
+
+/// Known-bits analysis.
+pub(super) struct KnownBits {
+    iidx_to_known_bits: IndexVec<InstIdx, Option<KnownBitValue>>,
```

Comment:
These are very long names, and not hugely helpful :) This could just be something like `known_bits` or even shorter.